File: Design Pattern/Structural/Proxy/databaseCache.cpp
#include <chrono>
#include <iostream>
#include <string>
#include <thread>
#include <unordered_map>

using namespace std;

class DatabaseService {
public:
  virtual string query(const string &sql) = 0;
  virtual ~DatabaseService() {}
};

class RealDatabaseService : public DatabaseService {
public:
  string query(const string &sql) override {
    cout << "RealDatabase: Executing query: " << sql << endl;
    this_thread::sleep_for(chrono::milliseconds(2000));
    return "Result for [" + sql + "]";
  }
};

class CachingDatabaseProxy : public DatabaseService {
private:
  RealDatabaseService realService;
  unordered_map<string, string> cache;

public:
  string query(const string &sql) override {
    auto it = cache.find(sql);
    if (it != cache.end()) {
      cout << "CachingProxy: Cache HIT for: " << sql << endl;
      return it->second;
    }
    cout << "CachingProxy: Cache MISS for: " << sql << endl;
    string result = realService.query(sql);
    cache[sql] = result;
    return result;
  }

  void clearCache() {
    cout << "CachingProxy: Cache cleared." << endl;
    cache.clear();
  }
};

int main() {
  CachingDatabaseProxy db;

  cout << "--- First query (cache miss) ---" << endl;
  cout << db.query("SELECT * FROM users") << endl;

  cout << "\n--- Same query again (cache hit) ---" << endl;
  cout << db.query("SELECT * FROM users") << endl;

  cout << "\n--- Different query (cache miss) ---" << endl;
  cout << db.query("SELECT * FROM orders WHERE status = 'pending'") << endl;

  cout << "\n--- Clear cache and retry ---" << endl;
  db.clearCache();
  cout << db.query("SELECT * FROM users") << endl;

  return 0;
}

File: Design Pattern/Structural/Proxy/databaseCache.ts
interface DatabaseService {
  query(sql: string): string;
}

class RealDatabaseService implements DatabaseService {
  query(sql: string): string {
    console.log("RealDatabase: Executing query: " + sql);
    const start = Date.now();
    while(Date.now() - start < 2000) {}
    return "Result for [" + sql + "]";
  }
}

class CachingDatabaseProxy implements DatabaseService {
  private realService = new RealDatabaseService();
  private cache: Map<string, string> = new Map();

  query(sql: string): string {
    const cached = this.cache.get(sql);
    if(cached !== undefined) {
      console.log("CachingProxy: Cache HIT for: " + sql);
      return cached;
    }

    console.log("CachingProxy: Cache MISS for: ", sql);
    const result = this.realService.query(sql);
    this.cache.set(sql, result);
    return result;
  }

  clearCache(): void {
    console.log("CachingProxy: Cache cleared.")
    this.cache.clear();
  }
}

const db = new CachingDatabaseProxy();

console.log("--- First query (cache miss) ---");
console.log(db.query("SELECT * FROM users"));

console.log("\n--- Same query again (cache hit) ---");
console.log(db.query("SELECT * FROM users"));

console.log("\n--- Different query (cache miss) ---");
console.log(db.query("SELECT * FROM orders WHERE status = 'pending'"));

console.log("\n--- Clear cache and retry ---");
db.clearCache();
console.log(db.query("SELECT * FROM users"));
